torch/_higher_order_ops/templated_attention.py

```diff
@@ -218,3 +248,437 @@ def templated_attention_fake_tensor_mode(
+    m = torch.arange(0, scores.size(2), device=scores.device)
+    n = torch.arange(0, scores.size(3), device=scores.device)
+
+    in_dim_buffers = (None,) * len(other_buffers)
+    score_mod = torch.vmap(fw_graph, in_dims=(0, None, None, None, 0) + in_dim_buffers)
+    score_mod = torch.vmap(score_mod, in_dims=(0, None, None, 0, None) + in_dim_buffers)
+    score_mod = torch.vmap(score_mod, in_dims=(0, None, 0, None, None) + in_dim_buffers)
+    score_mod = torch.vmap(score_mod, in_dims=(0, 0, None, None, None) + in_dim_buffers)
+
+    scores = score_mod(scores, b, h, m, n, *other_buffers).to(working_precision)
+
+    softmax_scores = torch.exp(scores - logsumexp.unsqueeze(-1))
+
+    grad_value = softmax_scores.to(query.dtype).transpose(-2, -1) @ grad_out
+
+    grad_softmax_scores = grad_out @ value.transpose(-2, -1)
+
+    sum_scores = torch.sum(out * grad_out, -1, keepdim=True)
+    grad_score_mod = softmax_scores * (grad_softmax_scores - sum_scores)
+
+    # Gradient of the inline score_mod function, with respect to the scores
+    in_dim_buffers = (None,) * len(other_buffers)
+    out_dims = [0, None, None, None, None]
+    joint_score_mod = torch.vmap(
+        joint_graph,
+        in_dims=(0, None, None, None, 0, 0) + in_dim_buffers,
+        out_dims=out_dims,
+    )
+    joint_score_mod = torch.vmap(
+        joint_score_mod,
+        in_dims=(0, None, None, 0, None, 0) + in_dim_buffers,
+        out_dims=out_dims,
+    )
+    joint_score_mod = torch.vmap(
+        joint_score_mod,
+        in_dims=(0, None, 0, None, None, 0) + in_dim_buffers,
+        out_dims=out_dims,
+    )
+    joint_score_mod = torch.vmap(
+        joint_score_mod,
+        in_dims=(0, 0, None, None, None, 0) + in_dim_buffers,
+        out_dims=out_dims,
+    )
+
+    grad_scores = joint_score_mod(scores, b, h, m, n, grad_score_mod, *other_buffers)[
+        0
+    ].to(query.dtype)
+
+    grad_query = grad_scores @ key
+    grad_key = grad_scores.transpose(-2, -1) @ query
+    return grad_query.contiguous(), grad_key.contiguous(), grad_value.contiguous()
+
+
+def trace_templated_attention_backward(
+    proxy_mode: ProxyTorchDispatchMode,
+    query: torch.Tensor,
+    key: torch.Tensor,
+    value: torch.Tensor,
+    out: torch.Tensor,
+    logsumexp: torch.Tensor,
+    grad_out: torch.Tensor,
+    fw_graph: Callable,  # GraphModule type hint?
+    joint_graph: Callable,
+    *other_buffers: torch.Tensor,
+) -> Tuple[torch.Tensor, torch.Tensor, torch.Tensor]:
+    """We already have the forward graph and joint graph from the forward pass, so we create a proxy attach both graphs"""
+    example_out = templated_attention_backward(
+        query,
+        key,
+        value,
+        out,
+        logsumexp,
+        grad_out,
+        fw_graph,
+        joint_graph,
+        *other_buffers,
+    )
+    # TODO If I dont wrap the graph modules in functionalize I already have my graph modules..
+    # Ask Brian
```

Comment:
@bdhirsh 
If I dont `ctx.functionalize(fw_graph)` and joint graph I think I will already have modules that were created in the autograd dispatch. This is done [here](https://github.com/pytorch/pytorch/pull/123902/files#diff-85e712896b6e0eb499cbfda5afd36d9fcf0c2988789cae78aeec8eb85c6fa18bR644-R645). I was thinking that since the fw_graph is already functional then the joint graph will be functional so I dont need that, and then would avoid having to remake the fx graphs here. But wasn't sure if I am breaking some invariant

Reply:
hmm... I think you're right that the "joint" graph you trace out for the score mod will get functionalized, so if you are stashing / re-using the backward part of it in this impl then it should be fine to use directly.

One problem though is that you still will need a functionalization rule for this backward op - it will need to unwrap all `FunctionalTensor` inputs and wrap outputs back into `FunctionalTensor`.

Reply:
offline discussion, leads me to believe its safer to leave as is